site/content/blog/2022-03-22-schema-migrations.md

```diff
@@ -0,0 +1,130 @@
+
+```Go
+type Locker interface {
+    Lock(context.Context) error
+    Unlock() error
+}
+```
+
+It essentially allows an instance to create a mutex value with expiration in the database and refreshes it until the instance finishes applying migrations. Other instances will keep waiting until the mutex is either removed from the database or get expired. The main difference here is that locking is not used for a single statement but it’s being used for the entire migration process. And this mechanism creates another level above sessions hence it wouldn’t fail if the advisory locks are bound to the sessions.
+
+The morph CLI provides the same functionality as the command line. Feel free to install and experiment as we think the command discovery should be sufficient to start using it!
+
+For those who have read this far, the following question may immediately come to mind: why you just didn't use an existing tool? Or at least, there should be some tools already developed for such problems (schema migrations). Yes, you are absolutely correct. Let us explain why we didn't use [golang-migrate](https://github.com/golang-migrate/migrate) which is one of the most popular projects and we already appreciate it. The main reason is a migration can fail at any time for various reasons, in a failure scenario `golang-migrate` leaves the migration state dirty and it requires some manual steps to fix the issue. Since we have idempotent migrations, it is OK to run a migration over and over again. So we don't need to use the dirty state. The application can even choose to roll back the migration in case of failure. The other motivation is to have a more granular schema version as we store every applied migration. Let's think of a scenario where you are planning to introduce 2 new migrations for the new release (let their version number be 18 and 19). At the last minute, you decided to pull out 18. migration. In this case, you need to rename 19. migration to 18 or rename the 18. migration to 20. but the way we store migrations do not require this. You can simply add 19. migration before the 18. one.
+
+## Summary
+
+Schema migrations enable versioning as the schema evolves. Making schema changes to databases is serious business as it means changing your customers' data structures. And that always comes with some risks. We consider it is a good practice to have easy recovery paths and to handle changes as a unit. We developed [morph](https://github.com/mattermost/morph/) to achieve these goals. It's being used on Channels product and we started to expand its usage on other products too. We are continuously improving it as requirements change and emerge. Feel free to try it out on your own application and we are always open to any type of contribution!
```

Comment:
Even though you technically linked it, perhaps add a more explicit call to action + link here somewhere (e.g. to replace the "Feel free to try it out" sentence) along the lines of. "If this seems like a tool that may benefit your project as well, we'd love you to give [morph](https://github.com/mattermost/morph/) a try and tell us what you think. The [README](https://github.com/mattermost/morph/blob/master/README.md) and [API docs](https://pkg.go.dev/github.com/mattermost/morph) should get you started, but if you run into any trouble feel free to create an issue, or reach out to us [on the Mattermost community server](https://community.mattermost.com/core/channels/developers-server)!

Reply:
Sure, makes sense.